api/src/main/java/org/apache/iceberg/expressions/AggregateUtil.java

```diff
@@ -0,0 +1,142 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import java.util.List;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/** Aggregate utility methods. */
+public class AggregateUtil {
+  private AggregateUtil() {}
+
+  /**
+   * Create a NestedField for this Aggregate Expression. This NestedField is used to build the
+   * pushed down aggregate schema.
+   *
+   * <p>e.g. SELECT COUNT(*), MAX(col1), MIN(col1), MAX(col2), MIN(col3) FROM table;
+   *
+   * <p>Suppose the table schema is Schema( Types.NestedField.required(1, "col1",
+   * Types.IntegerType.get()), Types.NestedField.required(2, "col2", Types.StringType.get()),
+   * Types.NestedField.required(3, "col3", Types.StringType.get()) );
+   *
+   * <p>The returned NestedField for the aggregates are Types.NestedField.required(1, COUNT(*),
+   * Types.LongType.get()) Types.NestedField.required(2, MAX(col1), Types.IntegerType.get())
+   * Types.NestedField.required(3, MIN(col1), Types.IntegerType.get()) Types.NestedField.required(4,
+   * MAX(col2), Types.StringType.get()) Types.NestedField.required(5, MIN(col3),
+   * Types.StringType.get())
+   */
+  public static Types.NestedField getAggregateNestedField(Expression expr, int index) {
+    return ExpressionVisitors.visit(expr, new AggregateNestedFieldVisitor(index));
+  }
+
+  /**
+   * Returns the column name this aggregate function is on. e.g. SELECT Max(col3) FROM table; This
+   * method returns col3
+   */
+  public static String getAggregateColumnName(Expression expr) {
+    return ExpressionVisitors.visit(expr, new AggregateColumnNameVisitor());
+  }
+
+  /**
+   * Returns the data type of this Aggregate Expression. The data type for COUNT is always Long. The
+   * data type for MAX and MIX are the same as the data type of the column this aggregate is applied
+   * on.
+   */
+  public static Type getAggregateType(Expression expr) {
+    return ExpressionVisitors.visit(expr, new AggregateTypeVisitor());
+  }
+
+  /**
+   * Returns the index of this Aggregate column in table schema. e.g. SELECT Max(col3) FROM table;
+   * Suppose the table has columns (col1, col2, col3), this method returns 2.
+   */
+  public static int columnIndexInTableSchema(Expression expr, Table table, boolean caseSensitive) {
+    return ExpressionVisitors.visit(expr, new AggregateColumnIndexVisitor(table, caseSensitive));
+  }
+
+  private static class AggregateNestedFieldVisitor
```

Comment:
I don't think that visitors are what you want for these methods. Visitors are used to transform trees. This appears to be trying to call a method using a visitor to cast the input. Since there are a lot of cases these visitors don't handle, their use is very narrow and probably not something you want to use a visitor for.

Reply:
I removed the visitors. 

I am not sure if I should implement this [eval](https://github.com/apache/iceberg/blob/898df4695e87dc615a631892a1529259ca61c10b/api/src/main/java/org/apache/iceberg/expressions/BoundAggregate.java#L31). It seems easier to get the values directly.

Reply:
The approach I took for this is to use `BoundAggregate.eval` to produce individual values from a row or data file. Then an `Aggregator` is used to aggregate those values. It seems to fit fairly cleanly.